components/script/dom/nodelist.rs

```diff
@@ -319,3 +327,39 @@ impl ChildrenList {
         self.last_index.set(0u32);
     }
 }
+
+// Labels lists: There might room for performance optimization
+// analogous to the ChildrenMutation case of a children list,
+// in which we can keep information from an older access live
+// if we know nothing has happened that would change it.
+// However, label relationships can happen from further away
+// in the DOM than parent-child relationships, so it's not as simple,
+// and it's possible that tracking label moves would end up no faster
+// than recalculating labels.
+#[derive(JSTraceable, MallocSizeOf)]
+#[unrooted_must_root_lint::must_root]
+pub struct LabelsList {
+    element: Dom<HTMLElement>,
+}
+
+impl LabelsList {
+    pub fn new(element: &HTMLElement) -> LabelsList {
+        LabelsList {
+            element: Dom::from_ref(element),
+        }
+    }
+
+    pub fn len(&self) -> u32 {
+        self.element.labels_vec().len() as u32
+    }
+
+    pub fn item(&self, index: u32) -> Option<DomRoot<Node>> {
+        let index = index as usize;
+        let vec = self.element.labels_vec();
```

Comment:
If the only time we use `labels_vec()` is to immediately index it, we should avoid spending the extra cost of iteration and allocation and have a `label_at(i)` function instead

Reply:
> If the only time we use `labels_vec()` is to immediately index it, we should avoid spending the extra cost of iteration and allocation and have a `label_at(i)` function instead

Yes. I realized that when I was working on an analogous case in #25428 and I should do that, assuming we are going this road and not the mutation observer one.

> Seems like Firefox uses a mutation observer to make the live list work efficiently. We do support them (I think), so this should be possible, but perhaps not necessary in this first pass.

#22264 is a separate attempt by @KiChjang at doing it that way, of which I was unaware when I worked on this. I don't see how that approach can be made to work in the case where the element and its label are in a detached subtree rather than part of a document, but I probably just don't know enough about mutation observers.

Reply:
Yeah we should land this with label_at, and then worry about mutation observers. Worth looking closer at Firefox's implementation to see what it does.

Reply:
Updated to separate label_at, labels_count methods.